fail to build framework for macos

##### System information (version)

- OpenCV => latest source code from this git repo
- Operating System / Platform => macOS 12.4
- Compiler => OpenCV python build_framework.py


##### Detailed description

build log:

```
 BUILD FAILED 


The following build commands failed:
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/CvTypeExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/DoubleVectorExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/FloatVectorExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/IntVectorExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/MatExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/Typealiases.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/dnn/DnnExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/face/FaceExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/features2d/Features2dExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/objdetect/ObjdetectExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/photo/PhotoExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/structured_light/Structured_lightExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/imgcodecs/ImgcodecsExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/imgproc/ImgprocExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/ml/MlExt.swift (in target 'opencv2' from project 'opencv2')
	EmitSwiftModule normal x86_64 (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/aruco/ArucoExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/calib3d/Calib3dExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/ByteVectorExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/core/CoreExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/text/TextExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/video/VideoExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/videoio/VideoioExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/wechat_qrcode/Wechat_qrcodeExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/xfeatures2d/Xfeatures2dExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwift normal x86_64 /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc_bindings_generator/osx/gen/objc/ximgproc/XimgprocExt.swift (in target 'opencv2' from project 'opencv2')
	CompileSwiftSources normal x86_64 com.apple.xcode.tools.swift.compiler (in target 'opencv2' from project 'opencv2')
	Copy /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/lib/Release/opencv2.framework/Versions/A/Headers/opencv2-Swift.h /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc/framework_build/opencv2.build/Release/opencv2.build/Objects-normal/x86_64/opencv2-Swift.h (in target 'opencv2' from project 'opencv2')
	Copy /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/lib/Release/opencv2.framework/Versions/A/Modules/opencv2.swiftmodule/x86_64-apple-macos.swiftdoc /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc/framework_build/opencv2.build/Release/opencv2.build/Objects-normal/x86_64/opencv2.swiftdoc (in target 'opencv2' from project 'opencv2')
	Copy /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/lib/Release/opencv2.framework/Versions/A/Modules/opencv2.swiftmodule/x86_64-apple-macos.swiftinterface /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc/framework_build/opencv2.build/Release/opencv2.build/Objects-normal/x86_64/opencv2.swiftinterface (in target 'opencv2' from project 'opencv2')
	Copy /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/lib/Release/opencv2.framework/Versions/A/Modules/opencv2.swiftmodule/x86_64-apple-macos.swiftmodule /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc/framework_build/opencv2.build/Release/opencv2.build/Objects-normal/x86_64/opencv2.swiftmodule (in target 'opencv2' from project 'opencv2')
	Copy /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/lib/Release/opencv2.framework/Versions/A/Modules/opencv2.swiftmodule/Project/x86_64-apple-macos.swiftsourceinfo /Users/andy/Workspace/OpenCV/build/build-x86_64-macosx/modules/objc/framework_build/opencv2.build/Release/opencv2.build/Objects-normal/x86_64/opencv2.swiftsourceinfo (in target 'opencv2' from project 'opencv2')
(32 failures)
============================================================
ERROR: Command '['xcodebuild', 'MACOSX_DEPLOYMENT_TARGET=10.12', 'ARCHS=x86_64', '-sdk', 'macosx', '-configuration', 'Release', '-parallelizeTargets', '-jobs', '8', '-target', 'ALL_BUILD', 'build']' returned non-zero exit status 65.
============================================================
Traceback (most recent call last):
  File "/Users/andy/Workspace/OpenCV/opencv/platforms/ios/build_framework.py", line 181, in build
    self._build(outdir)
  File "/Users/andy/Workspace/OpenCV/opencv/platforms/ios/build_framework.py", line 139, in _build
    self.buildOne(target[0], target[1], main_build_dir, cmake_flags)
  File "/Users/andy/Workspace/OpenCV/opencv/platforms/ios/build_framework.py", line 329, in buildOne
    execute(buildcmd + ["-target", "ALL_BUILD", "build"], cwd = builddir + "/modules/objc/framework_build")
  File "/Users/andy/Workspace/OpenCV/opencv/platforms/apple/cv_build_utils.py", line 13, in execute
    retcode = check_call(cmd, cwd = cwd)
  File "/Users/andy/opt/anaconda3/lib/python3.8/subprocess.py", line 364, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['xcodebuild', 'MACOSX_DEPLOYMENT_TARGET=10.12', 'ARCHS=x86_64', '-sdk', 'macosx', '-configuration', 'Release', '-parallelizeTargets', '-jobs', '8', '-target', 'ALL_BUILD', 'build']' returned non-zero exit status 65.
```


##### Steps to reproduce

1. follow this guide https://docs.opencv.org/4.x/d5/da3/tutorial_ios_install.html
2. `python opencv/platforms/osx/build_framework.py --macos_archs x86_64,arm64 --contrib opencv_contrib ./`

--- a/modules/objc/generator/gen_objc.py
+++ b/modules/objc/generator/gen_objc.py
@@ -774,30 +774,32 @@ class ObjectiveCWrapperGenerator(object):
         logging.info('ok: class %s, name: %s, base: %s', classinfo, name, classinfo.base)
         return classinfo
 
-    const_rename_map = dict()
-
     def add_const(self, decl, scope=None, enumType=None): # [ "const cname", val, [], [] ]
         constinfo = ConstInfo(decl, namespaces=self.namespaces, enumType=enumType)
         if constinfo.isIgnored():
             logging.info('ignored: %s', constinfo)
         else:
             objc_type = enumType.rsplit(".", 1)[-1] if enumType else ""
-            if constinfo.classname in const_fix and objc_type in const_fix[constinfo.classname] and constinfo.name in const_fix[constinfo.classname][objc_type]:
-                fixed_const = const_fix[constinfo.classname][objc_type][constinfo.name]
-                constinfo.name = fixed_const
-                constinfo.cname = fixed_const
+            if constinfo.enumType and constinfo.classpath:
+                new_name = constinfo.classname + '_' + constinfo.name
+                const_fix.setdefault(constinfo.classpath, {}).setdefault(objc_type, {})[constinfo.name] = new_name
+                constinfo.name = new_name
+                logging.info('use outer class prefix: %s', constinfo)
+
+            if constinfo.classpath in const_fix and objc_type in const_fix[constinfo.classpath]:
+                fixed_consts = const_fix[constinfo.classpath][objc_type]
+                if constinfo.name in fixed_consts:
+                    fixed_const = fixed_consts[constinfo.name]
+                    constinfo.name = fixed_const
+                    constinfo.cname = fixed_const
+                if constinfo.value in fixed_consts:
+                    constinfo.value = fixed_consts[constinfo.value]
 
             if not self.isWrapped(constinfo.classname):
                 logging.info('class not found: %s', constinfo)
-                constinfo.name = constinfo.classname + '_' + constinfo.name
+                if not constinfo.name.startswith(constinfo.classname + "_"):
+                    constinfo.name = constinfo.classname + '_' + constinfo.name
                 constinfo.classname = ''
-            elif constinfo.enumType and constinfo.classpath:
-                new_name = constinfo.classname + '_' + constinfo.name
-                self.const_rename_map[constinfo.name] = new_name
-                constinfo.name = new_name
-                if constinfo.value in self.const_rename_map:
-                    constinfo.value = self.const_rename_map[constinfo.value]
-                logging.info('use outer class prefix: %s', constinfo)
 
             ci = self.getClass(constinfo.classname)
             duplicate = ci.getConst(constinfo.name)

--- a/modules/objc/generator/gen_objc.py
+++ b/modules/objc/generator/gen_objc.py
@@ -789,6 +789,9 @@ class ObjectiveCWrapperGenerator(object):
                 logging.info('class not found: %s', constinfo)
                 constinfo.name = constinfo.classname + '_' + constinfo.name
                 constinfo.classname = ''
+            elif constinfo.enumType and constinfo.classpath:
+                constinfo.name = constinfo.classname + '_' + constinfo.name
+                logging.info('use outer class prefix: %s', constinfo)
 
             ci = self.getClass(constinfo.classname)
             duplicate = ci.getConst(constinfo.name)

--- a/modules/objc/generator/gen_objc.py
+++ b/modules/objc/generator/gen_objc.py
@@ -774,6 +774,8 @@ class ObjectiveCWrapperGenerator(object):
         logging.info('ok: class %s, name: %s, base: %s', classinfo, name, classinfo.base)
         return classinfo
 
+    const_rename_map = dict()
+
     def add_const(self, decl, scope=None, enumType=None): # [ "const cname", val, [], [] ]
         constinfo = ConstInfo(decl, namespaces=self.namespaces, enumType=enumType)
         if constinfo.isIgnored():
@@ -790,7 +792,11 @@ class ObjectiveCWrapperGenerator(object):
                 constinfo.name = constinfo.classname + '_' + constinfo.name
                 constinfo.classname = ''
             elif constinfo.enumType and constinfo.classpath:
-                constinfo.name = constinfo.classname + '_' + constinfo.name
+                new_name = constinfo.classname + '_' + constinfo.name
+                self.const_rename_map[constinfo.name] = new_name
+                constinfo.name = new_name
+                if constinfo.value in self.const_rename_map:
+                    constinfo.value = self.const_rename_map[constinfo.value]
                 logging.info('use outer class prefix: %s', constinfo)
 
             ci = self.getClass(constinfo.classname)

--- a/modules/objc/generator/gen_objc.py
+++ b/modules/objc/generator/gen_objc.py
@@ -203,6 +203,7 @@ class ConstInfo(GeneralInfo):
     def __init__(self, decl, addedManually=False, namespaces=[], enumType=None):
         GeneralInfo.__init__(self, "const", decl, namespaces)
         self.cname = get_cname(self.name)
+        self.swift_name = None
         self.value = decl[1]
         self.enumType = enumType
         self.addedManually = addedManually
@@ -783,6 +784,7 @@ class ObjectiveCWrapperGenerator(object):
             if constinfo.enumType and constinfo.classpath:
                 new_name = constinfo.classname + '_' + constinfo.name
                 const_fix.setdefault(constinfo.classpath, {}).setdefault(objc_type, {})[constinfo.name] = new_name
+                constinfo.swift_name = constinfo.name
                 constinfo.name = new_name
                 logging.info('use outer class prefix: %s', constinfo)
 
@@ -798,6 +800,7 @@ class ObjectiveCWrapperGenerator(object):
             if not self.isWrapped(constinfo.classname):
                 logging.info('class not found: %s', constinfo)
                 if not constinfo.name.startswith(constinfo.classname + "_"):
+                    constinfo.swift_name = constinfo.name
                     constinfo.name = constinfo.classname + '_' + constinfo.name
                 constinfo.classname = ''
 
@@ -1305,7 +1308,9 @@ $unrefined_call$epilogue$ret
                     ci.enum_declarations.write("""
 // C++: enum {1} ({2})
 typedef NS_ENUM(int, {1}) {{
-    {0}\n}};\n\n""".format(",\n    ".join(["%s = %s" % (c.name, c.value) for c in consts]), typeNameShort, typeName)
+    {0}\n}};\n\n""".format(
+                        ",\n    ".join(["%s = %s" % (c.name + (" NS_SWIFT_NAME(" + c.swift_name + ")" if c.swift_name else ""), c.value) for c in consts]),
+                        typeNameShort, typeName)
                     )
                 else:
                     if not wrote_consts_pragma:

